### ISSUE TITLE
Add rule to detect useless destructuring

### ISSUE THREAD

**Issue by ajafff** [fimbullinter/wotan#325]
Detect BindingPatterns that don't declare a variable and destructuring assignments that don't actually assign anything:

```ts
let {} = {};
({} = {});
({prop: {}} = {prop: {}});
({...{}} = {});

let [] = [];
[] = [];
[{}, ([, {nested: {}}])] = [{}, [{nested: {}}, {nested: {}}]];
[, , , ...[, [], ...[]]] = new Array(10);

function obj({}: Record<string, string>){}
function arr([, , , ]: any[]){}
```

The initial implementation doesn't need autofixing, because that's a bit more complicated if there are initializers with side effects.

**Comment by ajafff** [comment]
The code in my last comment is actually invalid. This syntax is not allowed, but TypeScript currently doesn't detect it: https://github.com/Microsoft/TypeScript/issues/26005

But this is still a thing for rest elements:

```ts
[...[a]] = [1];
// could be simplified to
[a] = [1];
```

**Comment by ajafff** [comment]
Also make sure to handle destructuring in FunctionRestParameter. Note that this is not yet allowed in TypeScript: https://github.com/Microsoft/TypeScript/issues/6275

```ts
function foo(...[]: any[]) {}
```

---

Maybe add an option to allow an empty destructuring pattern for parameters, as this is a good way to avoid unused parameter names:

```ts
function foo({}: string, []: string[], c: number, ...[]: any[]) { console.log(c); }
// instead of
function foo(_0: string, _1: string[], c: number, ..._rest: any[]) { console.log(c); }
```
